File: Design Pattern/Behavioral/Mediator/chat_mediator.ts
// Mediator interface
interface ChatMediator {
    sendMessage(message: string, sender: ChatUser): void;
    addUser(user: ChatUser): void;
}

class ChatRoom implements ChatMediator {
    private users: ChatUser[] = [];

    addUser(user: ChatUser): void {
        this.users.push(user);
        console.log(user.name + " joined the chat.");
    }

    sendMessage(message: string, sender: ChatUser): void {
        for (const user of this.users) {
            if (user !== sender) {
                user.receive(message, sender.name);
            }
        }
    }
}

class ChatUser {
    private mediator: ChatMediator;
    public name: string;

    constructor(mediator: ChatMediator, name: string) {
        this.mediator = mediator;
        this.name = name;
    }

    send(message: string): void {
        console.log(this.name + " sends: " + message);
        this.mediator.sendMessage(message, this);
    }

    receive(message: string, from: string): void {
        console.log(this.name + " receives from " + from + ": " + message);
    }
}

const chatRoom = new ChatRoom();

const alice = new ChatUser(chatRoom, "Alice");
const bob = new ChatUser(chatRoom, "Bob");
const charlie = new ChatUser(chatRoom, "Charlie");

chatRoom.addUser(alice);
chatRoom.addUser(bob);
chatRoom.addUser(charlie);

console.log();
alice.send("Hey everyone!");
console.log();
bob.send("Hi Alice!");

File: Design Pattern/Behavioral/Mediator/chat_mediator.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

class User;

class ChatMediator {
public:
  virtual void sendMessage(const string &message, User *sender) = 0;
  virtual void addUser(User *user) = 0;
  virtual ~ChatMediator() {}
};

class User {
protected:
  ChatMediator *mediator;
  string name;

public:
  User(ChatMediator *mediator, const string &name)
      : mediator(mediator), name(name) {}

  virtual void send(const string &message) = 0;
  virtual void receive(const string &message, const string &from) = 0;
  string getName() const { return name; }
  virtual ~User() {}
};

class ChatRoom : public ChatMediator {
private:
  vector<User *> users;

public:
  void addUser(User *user) override {
    users.push_back(user);
    cout << user->getName() << " joined the chat." << endl;
  }

  void sendMessage(const string &message, User *sender) override {
    for (User *user : users) {
      if (user != sender) {
        user->receive(message, sender->getName());
      }
    }
  }
};

class ChatUser : public User {
public:
  ChatUser(ChatMediator *mediator, const string &name) : User(mediator, name) {}

  void send(const string &message) override {
    cout << name << " sends: " << message << endl;
    mediator->sendMessage(message, this);
  }

  void receive(const string &message, const string &from) override {
    cout << name << " receives from " << from << ": " << message << endl;
  }
};

int main() {
  ChatRoom chatRoom;

  ChatUser alice(&chatRoom, "Alice");
  ChatUser bob(&chatRoom, "Bob");
  ChatUser charlie(&chatRoom, "Charlie");

  chatRoom.addUser(&alice);
  chatRoom.addUser(&bob);
  chatRoom.addUser(&charlie);

  cout << endl;
  alice.send("Hey everyone!");
  cout << endl;
  bob.send("Hi Alice!");

  return 0;
}
